File: data-structures/Binary Search Tree/C++/code.cpp
#include <iostream>

class BinarySearchTree
{
private:
    struct Node
    {
        int value;
        Node* left;
        Node* right;

        Node(int value)
            : value(value), left(nullptr), right(nullptr)
        {
        }
    };

    Node* root;

    Node* insert(Node* node, int value)
    {
        if (node == nullptr)
        {
            return new Node(value);
        }

        if (value < node->value)
        {
            node->left = insert(node->left, value);
        }
        else if (value > node->value)
        {
            node->right = insert(node->right, value);
        }

        return node;
    }

    Node* search(Node* node, int value) const
    {
        if (node == nullptr || node->value == value)
        {
            return node;
        }

        if (value < node->value)
        {
            return search(node->left, value);
        }

        return search(node->right, value);
    }

    Node* findMin(Node* node) const
    {
        Node* current = node;

        while (current != nullptr && current->left != nullptr)
        {
            current = current->left;
        }

        return current;
    }

    Node* remove(Node* node, int value)
    {
        if (node == nullptr)
        {
            return nullptr;
        }

        if (value < node->value)
        {
            node->left = remove(node->left, value);
        }
        else if (value > node->value)
        {
            node->right = remove(node->right, value);
        }
        else
        {
            // No children.
            if (node->left == nullptr && node->right == nullptr)
            {
                delete node;
                return nullptr;
            }

            // Only right child.
            if (node->left == nullptr)
            {
                Node* child = node->right;

                delete node;

                return child;
            }

            // Only left child.
            if (node->right == nullptr)
            {
                Node* child = node->left;

                delete node;

                return child;
            }

            // Two children.
            Node* successor = findMin(node->right);

            node->value = successor->value;
            node->right = remove(node->right, successor->value);
        }

        return node;
    }

    void inorder(Node* node) const
    {
        if (node == nullptr)
        {
            return;
        }

        inorder(node->left);
        std::cout << node->value << " ";
        inorder(node->right);
    }

    void preorder(Node* node) const
    {
        if (node == nullptr)
        {
            return;
        }

        std::cout << node->value << " ";
        preorder(node->left);
        preorder(node->right);
    }

    void postorder(Node* node) const
    {
        if (node == nullptr)
        {
            return;
        }

        postorder(node->left);
        postorder(node->right);
        std::cout << node->value << " ";
    }

    void freeTree(Node* node)
    {
        if (node == nullptr)
        {
            return;
        }

        freeTree(node->left);
        freeTree(node->right);

        delete node;
    }

public:
    BinarySearchTree()
        : root(nullptr)
    {
    }

    ~BinarySearchTree()
    {
        freeTree(root);
    }

    void insert(int value)
    {
        root = insert(root, value);
    }

    bool search(int value) const
    {
        return search(root, value) != nullptr;
    }

    void remove(int value)
    {
        root = remove(root, value);
    }

    void inorder() const
    {
        inorder(root);
        std::cout << "\n";
    }

    void preorder() const
    {
        preorder(root);
        std::cout << "\n";
    }

    void postorder() const
    {
        postorder(root);
        std::cout << "\n";
    }
};

int main()
{
    BinarySearchTree tree;

    tree.insert(50);
    tree.insert(30);
    tree.insert(70);
    tree.insert(20);
    tree.insert(40);
    tree.insert(60);
    tree.insert(80);

    std::cout << "In-order: ";
    tree.inorder();

    std::cout << "Pre-order: ";
    tree.preorder();

    std::cout << "Post-order: ";
    tree.postorder();

    std::cout << "Search 40: "
              << (tree.search(40) ? "found" : "not found")
              << "\n";

    tree.remove(30);

    std::cout << "After deleting 30: ";
    tree.inorder();

    return 0;
}

File: data-structures/Binary Search Tree/Python/code.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self, value):
        self.root = self._insert(self.root, value)

    def _insert(self, node, value):
        if node is None:
            return Node(value)

        if value < node.value:
            node.left = self._insert(node.left, value)
        elif value > node.value:
            node.right = self._insert(node.right, value)

        return node

    def search(self, value):
        return self._search(self.root, value)

    def _search(self, node, value):
        if node is None:
            return False

        if node.value == value:
            return True

        if value < node.value:
            return self._search(node.left, value)

        return self._search(node.right, value)

    def delete(self, value):
        self.root = self._delete(self.root, value)

    def _delete(self, node, value):
        if node is None:
            return None

        if value < node.value:
            node.left = self._delete(node.left, value)

        elif value > node.value:
            node.right = self._delete(node.right, value)

        else:
            # No children.
            if node.left is None and node.right is None:
                return None

            # Only right child.
            if node.left is None:
                return node.right

            # Only left child.
            if node.right is None:
                return node.left

            # Two children.
            successor = self._find_min(node.right)

            node.value = successor.value
            node.right = self._delete(
                node.right,
                successor.value
            )

        return node

    def _find_min(self, node):
        current = node

        while current.left is not None:
            current = current.left

        return current

    def inorder(self):
        result = []
        self._inorder(self.root, result)
        print(*result)

    def _inorder(self, node, result):
        if node is None:
            return

        self._inorder(node.left, result)
        result.append(node.value)
        self._inorder(node.right, result)

    def preorder(self):
        result = []
        self._preorder(self.root, result)
        print(*result)

    def _preorder(self, node, result):
        if node is None:
            return

        result.append(node.value)
        self._preorder(node.left, result)
        self._preorder(node.right, result)

    def postorder(self):
        result = []
        self._postorder(self.root, result)
        print(*result)

    def _postorder(self, node, result):
        if node is None:
            return

        self._postorder(node.left, result)
        self._postorder(node.right, result)
        result.append(node.value)


tree = BinarySearchTree()

tree.insert(50)
tree.insert(30)
tree.insert(70)
tree.insert(20)
tree.insert(40)
tree.insert(60)
tree.insert(80)

print("In-order:", end=" ")
tree.inorder()

print("Pre-order:", end=" ")
tree.preorder()

print("Post-order:", end=" ")
tree.postorder()

print(
    "Search 40:",
    "found" if tree.search(40) else "not found"
)

tree.delete(30)

print("After deleting 30:", end=" ")
tree.inorder()